File: DownloadFiles/TGNet/Chapter04/VB/Exercise4-5/Debug Version/MileageCalculator/MileageCalculator/Form1.vb
Public Class Form1

  Private Sub btnGetMileage_Click _
  (ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnGetMileage.Click
    'Call the GetAnswer Method
    GetAnswer()
    MessageBox.Show("Done")
  End Sub

  'Create a Method to calculate Mileage
  Public Sub GetAnswer()

    'Create 3 variables to work with
    Dim Mileage, Distance, Gallons As Decimal

    Try 'Add this line to start Try-Catch Block
      ' Set values from the UI
      Distance = txtDistance.Text
      Gallons = txtGallons.Text

      'Check for zeros
      If Distance = 0 Then
        'Cause an Exception to be Thrown
        Throw New MyCustomException
      End If

      'Calculate the mileage
      Mileage = Distance / Gallons

    Catch ex As DivideByZeroException
      'Catch Divide by Zero errors and use the object from Microsoft's  
      'DivideByZeroException Class to show a Message 
      MessageBox.Show(ex.Message)

    Catch ex As MyCustomException
      'Catch all MyCustomException errors, 
      'Use the Object to show a Custom error Message
      'MessageBox.Show(ex.Message) < -- This line is no longer needed

      'Step 1: Create the TraceListener Objects
      Dim objTWFile As New TextWriterTraceListener("c:\Trace.log")
      Dim objTWCon As New TextWriterTraceListener(Console.Out)
      Dim objEL As New EventLogTraceListener("Mileage Calculator")

      'Step 2: Add the Objects to the Listeners collection
      Trace.Listeners.Add(objTWFile)
      Trace.Listeners.Add(objTWCon)
      Trace.Listeners.Add(objEL)

      'Step 3: Set the system to automatic mode
      Trace.AutoFlush = True

      'Step 4: Set the message you want to send
      Dim strMessage As String
      strMessage = "======( Trace Info: " + Now() + " )======" + vbCrLf
      strMessage += ex.Message + vbCrLf
      strMessage += ex.ToString()

      'Step 5: Send the message to all of the listeners, including the default
      Trace.WriteLine(strMessage)

      'Step 6: Close the file in the file version and clear all the Listeners 
      objTWFile.Close()
      Trace.Listeners.Clear()

    Catch ex As Exception
      'Catch any other type of Exception
      MessageBox.Show("There was an Error")

    Finally
      'Put code here that you want to always run
      MessageBox.Show("Inside Finally")
    End Try
    MessageBox.Show(Mileage.ToString, "Answer")

  End Sub

End Class

File: DownloadFiles/TGNet/Chapter04/C#/Exercise4-5/Debug Version/MileageCalculator/MileageCalculator/Form1.cs
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using System.Diagnostics;

#endregion

namespace MileageCalculator
{
  partial class Form1 : Form
  {
      public Form1()
      {
          InitializeComponent();
      }

      private void btnGetMileage_Click(object sender, EventArgs e)
      {
          GetAnswer();
          MessageBox.Show("Done");
      }
  
     //Create a Method to calculate Mileage
     public void  GetAnswer()
      {
        //Create 3 variables to work with
        Decimal Mileage = 0, Distance = 0, Gallons = 0;

        try
        {
          //Add code to set these values from the UI
          Distance = Convert.ToDecimal(txtDistance.Text);
          Gallons = Convert.ToDecimal(txtGallons.Text);

          //Check for zeros
          if (Distance == 0)
          {
            //Cause an Exception to be Thrown
            throw new MyCustomException();
          }

          Mileage = Distance / Gallons;

        }
        catch (DivideByZeroException ex)
        {
          //Catch Divide by Zero errors and use the object from Microsoft's  
          //DivideByZeroException Class to show a Message 
          MessageBox.Show(ex.Message); 
        }
catch (MyCustomException ex)
{
  //Catch all MyCustomException errors, 
  //Use the Object to show a Custom error Message
  //MessageBox.Show(ex.Message); <-- This line is no longer needed

  //Step 1: Create the TraceListener Objects
  TextWriterTraceListener objTWCon = new TextWriterTraceListener(Console.Out);
  TextWriterTraceListener objTWFile = new TextWriterTraceListener("c:\\Trace.log");
  EventLogTraceListener objEL = new EventLogTraceListener("Mileage Calculator");

  //Step 2: Add the Objects to the Listeners collection
  Debug.Listeners.Add(objTWFile);
  Debug.Listeners.Add(objTWCon);
  Debug.Listeners.Add(objEL);

  //Step 3: Set the system to automatic mode
  Debug.AutoFlush = true;

  //Step 4: Set the message you want to send
  string strMessage;
  strMessage = "======( Trace Info: " + DateTime.Now + " )======" + "\n";
  strMessage += ex.Message + "\n";
  strMessage += ex.ToString();

  //Step 5: Send the message to all of the listeners, including the default
  Debug.WriteLine(strMessage);

  //Step 6: Close the file in the file version and clear all the Listeners 
  objTWFile.Close();
  Debug.Listeners.Clear();

}
        catch
        {
          //Catch any other type of Exception
          MessageBox.Show("There was an Error");
        }
        finally
        {
          //Put code here that you want to always run 
          MessageBox.Show("Inside Finally");
        } //End Try-Catch Block

        MessageBox.Show(Mileage.ToString(), "Answer");

       }//end of GetAnswer

  } //End of Class
}